doc/gen_docs.go

```diff
@@ -0,0 +1,170 @@
+	ChildrenLinks []string // rendered internal links to the child commands as a slice
+	RelatedLinks  []string // rendered internal links to the related commands as a slice
+	CommandLink   string   // rendered internal link to the command
+	HeaderScale   int      // integer scale indicating depth of the current command
+	AutoGenTag    string   // automatically generated tag by Cobra
+}
+
+func generateCmdOutline(cmd *cobra.Command, linkHandler func(string) string) *CmdOutline {
+	name := cmd.CommandPath()
+	short := cmd.Short
+	long := cmd.Long
+	if len(long) == 0 {
+		long = short
+	}
+
+	useLine := cmd.UseLine()
+
+	example := cmd.Example
+
+	buf := new(bytes.Buffer)
+
+	var flagString string
+	flags := cmd.NonInheritedFlags()
+	flags.SetOutput(buf)
+	if flags.HasAvailableFlags() {
+		flags.PrintDefaults()
+		flagString = buf.String()
+		buf.Reset()
+	}
+
+	var flagSlice []string
+	flagSplitter := regexp.MustCompile("\n *-") // clips first '-' in flags starting the second flag
+	numberOfFlags := strings.Count(flagString, "--")
+	flagSlice = flagSplitter.Split(flagString, numberOfFlags)
+	for i, flag := range flagSlice {
+		if i > 0 {
+			flagSlice[i] = "-" + flag // add clipped '-' back to flags
+		}
+	}
+
+	var parentFlagString string
+	parentFlags := cmd.InheritedFlags()
+	parentFlags.SetOutput(buf)
+	if parentFlags.HasAvailableFlags() {
+		parentFlags.PrintDefaults()
+		parentFlagString = buf.String()
+		buf.Reset()
+	}
+
+	headerScale := 0
+	var parentLink string
+	if cmd.HasParent() {
+		parent := cmd.Parent()
+		pname := parent.CommandPath()
+		link := pname + ".md"
+		link = strings.Replace(link, " ", "_", -1)
+		parentLink = fmt.Sprintf("* [%s](%s)\t - %s\n", pname, linkHandler(link), parent.Short)
+
+		headerScale = 1
+		for parent.HasParent() {
+			headerScale += 1
+			parent = parent.Parent()
+		}
+	}
+
+	var childrenLinks []string
+	children := cmd.Commands()
+	sort.Sort(byName(children))
+
+	for _, child := range children {
+		var childLink string
+		if !child.IsAvailableCommand() || child.IsAdditionalHelpTopicCommand() {
+			continue
+		}
+		cname := name + " " + child.Name()
+		link := cname + ".md"
+		link = strings.Replace(link, " ", "_", -1)
+		childLink = fmt.Sprintf("* [%s](%s)\t - %s\n", cname, linkHandler(link), child.Short)
+		childrenLinks = append(childrenLinks, childLink)
+	}
+
+	var relatedLinks []string
+	relatedCmds := cmd.RelatedCommands()
+
+	for _, relCmd := range relatedCmds {
+		var relatedLink string
+		if !relCmd.IsAvailableCommand() || relCmd.IsAdditionalHelpTopicCommand() {
+			continue
+		}
+		rname := relCmd.CommandPath()
+		link := rname + ".md"
+		link = strings.Replace(link, " ", "_", -1)
+		relatedLink = fmt.Sprintf("* [%s](%s)\t - %s\n", rname, linkHandler(link), relCmd.Short)
+		relatedLinks = append(relatedLinks, relatedLink)
+	}
+
+	var commandLink string
+	link := name + ".md"
+	link = strings.Replace(link, " ", "_", -1)
+	commandLink = fmt.Sprintf("%s", linkHandler(link))
```

Comment:
Is this the right place to put link generation in? Everything else is pretty markdown agnostic, but not this. Wouldn't `linkHandler` be the function that generates the link? Otherwise, what is it's purpose?

Regardless, 

```
commandLink = fmt.Sprintf("%s", linkHandler(link))
```

can just be

```
commandLink = linkHandler(link)
```

Reply:
The original markdown generator in Cobra generates links of these forms, and `linkHandler` acts as a function for users to manually adjust the link provided to their liking.

Reply:
Okay I thought about it, and I think in order to maintain a similar principle as what Cobra was originally doing, we should define a default link. 
```
link := name + ".md"
link = strings.Replace(link, " ", "_", -1)
```
is part of the original code, so I want to make sure that that still exists for the markdown generator. If we follow that same kind of thing, then it made sense to me for there to be a default "link" for custom templates (granted, I don't know how this would be an effective link at all). Alternatively, these links could not exist for custom templates and they could be blank. Not too sure what's best there...

Either way, I would like to keep the purpose of the `linkHandler` consistent with upstream cobra, which is to customize the default rendered links.